### PR TITLE
Undeprecate lint `unstable_features` and make use of it in the compiler

### PR DESCRIPTION
**Pull request by fmease**
See also #117937.

r? compiler

### CODE REVIEW

**Comment by fmease** [comment]
TODO: Write up.

Nominating for lang team discussion.

File: compiler/rustc_lint/src/builtin.rs [hunk 1]
```diff
@@ -1232,10 +1232,26 @@ impl<'tcx> LateLintPass<'tcx> for MutableTransmutes {
...
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Crates that use unstable features require a nightly compiler which might not be desirable.
+    /// Especially in larger projects with a multitude of crates where some crates are allowed to
+    /// use unstable features and some are not, this lint may come in handy to enforce policies of
+    /// this kind.
```
Comment (Nilstrieb) [inline compiler/rustc_lint/src/builtin.rs:1251]: maybe add an explicit note that this is only necessarily when the project is normally built using a nightly compiler and irrelevant for anyone using the stable compiler?

**Comment by compiler-errors** [comment]
This doesn't fully fix #117937 because `rustc_type_ir` *must* build both with and without its `feature = "nightly"`. There's no way to just add `#![deny(unstable_features)]` to `rustc_type_ir` without both a major perf and ergonomics hit to rustc.

**Comment by fmease** [comment]
> This doesn't fully fix #117937 because rustc_type_ir must build both with and without its feature = "nightly".

Do you mean that you'd like to see CI building rustc_type_ir twice? Once with and once without `RUSTC_BOOTSTRAP=1` (and maybe have `#![cfg_attr(not(feature = "nightly"), deny(unstable_feature))]`)? I can remove the "Fixes" from the description.

For the lexer and the other crate, I feel `deny(unstable_features)` is a perfect fit.

**Comment by compiler-errors** [comment]
> Do you mean that you'd like to see CI building rustc_type_ir twice? Once with and once without RUSTC_BOOTSTRAP=1

Yeah, that would be ideal.

> For the lexer and the other crate, I feel deny(unstable_features) is a perfect fit.

Yes, I agree.

**Comment by WaffleLapkin** [comment]
(I'm removing https://github.com/rust-lang/rust/labels/S-waiting-on-review so it doesn't show on my todo list, feel free to add the label back once t-lang did t-lang things)

**Comment by tmandry** [comment]
@rfcbot merge

**Comment by fmease** [comment]
force-push: doc-only change to apply Nils's suggestion and to integrate a point made by errs in the t-lang triag meeting

**Comment by traviscross** [comment]
We discussed this in the 2023-12-20 T-lang triage meeting.  The consensus was positive on doing this, resulting in the proposed FCP merge above.  Thanks to @fmease for providing context in the meeting and pushing this forward.

**Comment by fmease** [comment]
@WaffleLapkin The FCP has finished, PR is now ready for further reviews. Take your time or reassign if you prefer :)

**Review by WaffleLapkin — APPROVED** [review]
Looks good to me :)

**Pull request merged**